### DISCUSSION TITLE
Negative WSS@95 value generated using uncertainty query strategy (in simulation mode)

### DISCUSSION THREAD

**Discussion by brittdev** [asreview/asreview#994] [Q&A]
Apologies if this belongs in Discussion rather than Issues but I feel like this might be a bug since I don't understand why it's not working. 

**Describe the bug**
When generating the statistics for the NB + TFIDF model combination using uncertainty as the query strategy, simulation mode showed a negative value for WSS@95. (Yes, I still use simulation mode) 
 
**To Reproduce**
The command that I used to generate the result files was: 
`asreview simulate-batch [my_dataset].csv --state_file [my_dataset]_nbtfidf_uncertainty.h5 --seed 1 --init_seed 1 -m nb -e tfidf -q uncertainty --n_run 10`

**Screenshots**
Couldn't fit it in a screenshot, so here is my output:

-----------  general  -----------
Number of runs            : 10
Number of papers          : 178
Number of included papers : 75
Number of excluded papers : 103
Number of unlabeled papers: 0
Number of queries         : 177

-----------  settings  -----------
data_name         : Robot_longlist
model             : nb
query_strategy    : uncertainty
balance_strategy  : double
feature_extraction: tfidf
n_instances       : 1
n_prior_included  : 1
n_prior_excluded  : 1
mode              : simulate
model_param       : {'alpha': 3.822}
query_param       : {}
feature_param     : {'ngram_max': 1, 'stop_words': 'english', 'split_ta': 0, 'use_keywords': 0}
balance_param     : {'a': 2.155, 'alpha': 0.94, 'b': 0.789, 'beta': 1.0}

-----------    ATD    -----------
 0.68

Time to discovery:

    row   : value
    162   : 28.8
    58    : 45.7
    111   : 46.0
    27    : 51.5
    117   : 58.666666666666664
    42    : 59.0
    153   : 64.8
    101   : 65.1
    65    : 70.0
    163   : 72.77777777777777
    7     : 80.3
    164   : 80.5
    138   : 81.9
    21    : 83.1
    144   : 87.6
    15    : 89.0
    169   : 89.5
    70    : 90.8
    41    : 95.55555555555556
    73    : 97.5
    93    : 101.9
    157   : 104.1
    82    : 106.5
    161   : 106.9
    114   : 109.5
    50    : 109.8
    35    : 110.5
    48    : 113.8
    98    : 115.8
    131   : 116.3
    119   : 116.6
    39    : 117.2
    167   : 119.9
    149   : 120.5
    19    : 122.4
    61    : 123.11111111111111
    106   : 124.7
    176   : 125.3
    26    : 126.1
    75    : 127.7
    165   : 127.8
    71    : 131.44444444444446
    103   : 135.6
    43    : 136.3
    1     : 136.5
    128   : 137.44444444444446
    23    : 138.22222222222223
    172   : 138.9
    44    : 140.4
    53    : 141.5
    150   : 143.5
    171   : 145.5
    129   : 146.6
    38    : 146.8
    14    : 148.5
    102   : 148.6
    59    : 151.33333333333334
    54    : 152.4
    133   : 152.4
    136   : 152.44444444444446
    37    : 153.8
    132   : 155.2
    80    : 157.0
    57    : 159.3
    16    : 159.72727272727272
    135   : 160.3
    126   : 163.8
    28    : 163.9
    141   : 167.1
    160   : 167.5
    3     : 169.5
    33    : 171.66666666666666
    121   : 172.6
    47    : 174.1
    12    : 177.3684210526316

-----------  WSS/RRF  -----------
WSS@95 : -2.35 %
WSS@100: 0.00  %
RRF@5  : 3.38  %
RRF@10 : 4.05  %

**Version information**
 - OS: Windows 10 
 - ASReview version 0.19

**Additional context**
Other datasets also returned negative WSS@95 values using the uncertainty query strategy. For one dataset that I tried, the random query strategy also returned a negative WSS@95 value for the NB+TFIDF model combination. 

Do these query strategies not work well together with the NB+TFIDF model combination? I haven't tried it yet for other model combinations. Or is this an issue with simulation mode perhaps? Or also possible, is there a typo in my prompt?

**Comment by Rensvandeschoot** [comment]
_Yes, I still use simulation mode_
The [simulation mode](https://asreview.nl/blog/simulation-mode-class-101/) is one of the most brilliant aspects of ASReview, so I can only recommend using it!!!

Have you tried to visualize the recall curve? It might help to see what is going on? See https://github.com/asreview/asreview-visualization

The uncertainty query strategy will select the most relevant records to train the model (e.g., close to a relevance score of .5) instead of presenting the most relevant ones, as with certainty-based sampling. So, you will not save any time by using this option, but you have trained a better model. This is the reason why the WSS@95 is low. 

I also noticed you have quite a large percentage of relevant documents (75 of 178). The system is optimized for finding a very small portion of relevant ones in a large pool of records, and 178 is not so many. 

So, the above combination might have caused the WSS to become negative. I don't think we programmed the WSS in such a way it is bounded to be between 0-1, right @J535D165 ?

**Comment by brittdev** [comment]
Thanks for the suggestion! I had not tried to plot the recall curve yet, but when I just did it looked like this: 

![image](https://user-images.githubusercontent.com/60607063/154274180-c2aed8a3-a3c5-4931-a2f0-fe96c92ca9b8.png)

So if I interpret this correctly the WSS@95 should be ~1,9%. 

> I also noticed you have quite a large percentage of relevant documents (75 of 178). The system is optimized for finding a very small portion of relevant ones in a large pool of records, and 178 is not so many.

I tried the uncertainty query strategy on a different dataset that contains 1716 records of which 63 records were labeled relevant. According to the generated statistics the WSS@95 was -4,31% and the recall curve looks like this: 

![image](https://user-images.githubusercontent.com/60607063/154275855-0318fca6-5445-4d98-9995-b9d309f7db83.png)

So the size of the dataset and inclusion rate also doesn't seem to be the problem here.

**Comment by Rensvandeschoot** [comment]
This plot makes perfect sense because you are not selecting the most relevant records, you are selecting the records most useful to the model with a relevance score close to .05. So, after reading 95% of the records, you did not gain anything from using active learning in terms of having seen the relevant records. 

Try to run a model with certainty based sampling (max) and compare the results...
